endpoints/bookstore-grpc/client/src/main/java/com/google/endpoints/examples/bookstore/BookstoreClient.java

```diff
@@ -0,0 +1,235 @@
+
+package com.google.endpoints.examples.bookstore;
+
+import com.google.protobuf.Empty;
+
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.ClientInterceptors;
+import io.grpc.ForwardingClientCall;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.CommandLineParser;
+import org.apache.commons.cli.DefaultParser;
+import org.apache.commons.cli.HelpFormatter;
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.apache.commons.cli.ParseException;
+
+/**
+ * A client application which calls the Bookstore API over gRPC.
+ */
+public final class BookstoreClient {
+
+  private static final String defaultBookstore = "localhost:8000";
+
+  public static void main(String[] args) throws Exception {
+    Options options = createOptions();
+    CommandLineParser parser = new DefaultParser();
+    CommandLine line;
+    try {
+      line = parser.parse(options, args);
+    } catch (ParseException e) {
+      System.err.println("Invalid command line: " + e.getMessage());
+      printUsage(options);
+      return;
+    }
+
+    String address = defaultBookstore;
+    String apiKey = null;
+    String authToken = null;
+    String operation = "list";
+
+    if (line.hasOption("bookstore")) {
+      address = line.getOptionValue("bookstore");
+    }
+
+    if (line.hasOption("api_key")) {
+      apiKey = line.getOptionValue("api_key");
+    }
+
+    if (line.hasOption("auth_token")) {
+      authToken = line.getOptionValue("auth_token");
+    }
+
+    if (line.hasOption("operation")) {
+      operation = line.getOptionValue("operation");
+    }
+
+    // Create gRPC stub.
+    BookstoreGrpc.BookstoreBlockingStub bookstore = createBookstoreStub(
+        address, apiKey, authToken);
+
+    if ("list".equals(operation)) {
+      listShelves(bookstore);
+    } else if ("create".equals(operation)) {
+      createShelf(bookstore);
+    } else if ("enumerate".equals(operation)) {
+      enumerate(bookstore);
+    }
+  }
+
+  private static final class Interceptor implements ClientInterceptor {
+    private final String apiKey;
+    private final String authToken;
+
+    private static Metadata.Key<String> apiKeyHeader =
+        Metadata.Key.of("x-api-key", Metadata.ASCII_STRING_MARSHALLER);
+    private static Metadata.Key<String> authorizationHeader =
+        Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER);
+
+    public Interceptor(String apiKey, String authToken) {
+      this.apiKey = apiKey;
+      this.authToken = authToken;
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+        MethodDescriptor<ReqT,RespT> method, CallOptions callOptions, Channel next) {
+      System.out.println("Intercepted " + method.getFullMethodName());
```

Comment:
This seems like it'd be better as either a log message (if it's meant to inform, while the program is running), or as a javadoc (if it's meant to educate about what this method is here for). Printing to stdout seems a bit noisy for an interception method.